### PR TITLE
merge `need_type_info_err(_const)`

### LINKED ISSUE #77092: Malformed error message with wrong use of min const generics
This is wrong code (because there's no way to infer N):

```rust
#![feature(min_const_generics)]

use std::convert::TryInto;

#[inline(never)]
fn take_array_from_mut<T, const N: usize>(data: &mut [T], start: usize) -> &mut [T; N] {
    (&mut data[start .. start + N]).try_into().unwrap()
}

fn main() {
    let mut arr = [0, 1, 2, 3, 4, 5, 6, 7, 8];

    for i in 1 .. 4 {
        println!("{:?}", take_array_from_mut(&mut arr, i));
    }
}
```

rustc `V. 1.48.0-nightly 0da580074 2020-09-22` gives:

```rust
error[E0282]: type annotations needed
  |
  = note: unable to infer the value of a const parameter

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.

Tool completed with exit code 1
```

### PR DESCRIPTION
**Pull request by lcnr**
I hoped that this would automatically solve #76737 but it doesn't quite seem like it

fixes #77092

r? @varkor

### CODE REVIEW

File: compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs [hunk 1]
```diff
@@ -220,60 +223,124 @@ impl Into<rustc_errors::DiagnosticId> for TypeAnnotationNeeded {
 impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
     pub fn extract_type_name(
```
Comment (varkor) [inline compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs:224]: This should be renamed, if it's no longer just for types.

File: compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs [hunk 2]
```diff
@@ -220,60 +223,124 @@ impl Into<rustc_errors::DiagnosticId> for TypeAnnotationNeeded {
...
         &self,
-        ty: Ty<'tcx>,
+        arg: GenericArg<'tcx>,
         highlight: Option<ty::print::RegionHighlightMode>,
     ) -> (String, Option<Span>, Cow<'static, str>, Option<String>, Option<&'static str>) {
```
Comment (varkor) [inline compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs:228]: I know this is a pre-existing issue, but could you extract this result value into a struct? It's not possible to guess what these components mean from the signature.

File: compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs [hunk 3]
```diff
@@ -656,18 +681,22 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
         );
-        err.span_label(span, InferCtxt::missing_type_msg(&name, &descr, parent_name, parent_descr));
+        err.span_label(
+            span,
+            InferCtxt::missing_type_msg("type", &name, &descr, parent_name, parent_descr),
+        );
         err
     }
 
     fn missing_type_msg(
```
Comment (varkor) [inline compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs:691]: This should be renamed too.

File: src/test/ui/const-generics/infer/issue-77092.stderr [hunk 4]
```diff
@@ -0,0 +1,9 @@
+error[E0282]: type annotations needed
+  --> $DIR/issue-77092.rs:13:26
+   |
+LL |         println!("{:?}", take_array_from_mut(&mut arr, i));
+   |                          ^^^^^^^^^^^^^^^^^^^ cannot infer the value of the constant `{_: usize}`
```
Comment (varkor) [inline src/test/ui/const-generics/infer/issue-77092.stderr:5]: Would it make more sense here to use a similar message to below?
E.g.
```
cannot infer the value of const parameter `N`
```

**Review by varkor — COMMENTED** [review]
I've just skimmed the implementation for now: I'll take a closer look later. But I'm really happy to see this code cleaned up!

File: src/test/ui/const-generics/infer/issue-77092.stderr (hunk 4 above)
Comment (lcnr) [reply src/test/ui/const-generics/infer/issue-77092.stderr:5]: the failed inference variable has a `ConstInference` origin, as it was created from unifying two uninferred consts vids.

So we don't have the info to emit a better error message in this case. We probably should "just" use one of the `ConstParam`s as the origin in that case. I think the reason we don't do so with types is due to variance.

Not completely sure about all of this though, so I would prefer to look into this in a followup PR

**Comment by lcnr** [comment]
Implement the name changes and put the result of previously `extract_type_name` into a new struct.

I am myself not completely happy with the new names but wasn't able to think of anything better :laughing:

File: compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs [hunk 5]
```diff
@@ -217,65 +220,151 @@ impl Into<rustc_errors::DiagnosticId> for TypeAnnotationNeeded {
...
+                }
+            }
+            GenericArgKind::Lifetime(_) => bug!("unexpected lifetime"),
         }
-        let _ = ty.print(printer);
-        (s, None, ty.prefix_string(), None, None)
     }
 
-    // FIXME(eddyb) generalize all of this to handle `ty::Const` inference variables as well.
     pub fn need_type_info_err(
```
Comment (varkor) [inline compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs:352]: Maybe something like `emit_inference_failure_err`?

File: compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs [hunk 6]
```diff
@@ -217,65 +220,151 @@ impl Into<rustc_errors::DiagnosticId> for TypeAnnotationNeeded {
...
+    pub description: Cow<'static, str>,
+    pub parent_name: Option<String>,
+    pub parent_description: Option<&'static str>,
+}
+
 impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
-    pub fn extract_type_name(
+    /// Extracts data used by diagnostic for either types or constants
+    /// which were stuck during inference.
+    pub fn extract_infer_data(
```
Comment (varkor) [inline compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs:235]: Maybe `extract_inference_diagnostics_data`?

File: compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs [hunk 7]
```diff
@@ -217,65 +220,151 @@ impl Into<rustc_errors::DiagnosticId> for TypeAnnotationNeeded {
     }
 }
 
+/// Information about a constant or a type containing inference variables.
+pub struct InferDiagnosticsData {
```
Comment (varkor) [inline compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs:224]: I slightly prefer `InferenceDiagnosticsData`, but I don't have a strong preference.

**Review by varkor — COMMENTED** [review]
Actual implementation looks good, so r=me after renaming `ty` and if you're happy with the names.

**Pull request merged**